src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -12,21 +11,24 @@
 
 namespace Microsoft.Data.SqlClient
 {
-
-    /// <summary>
-    /// Default auth provider for AD Integrated.
-    /// </summary>
-    internal class ActiveDirectoryAuthenticationProvider : SqlAuthenticationProvider
+    /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ActiveDirectoryAuthenticationProvider/*'/>
+    public class ActiveDirectoryAuthenticationProvider : SqlAuthenticationProvider
     {
         private static readonly string s_defaultScopeSuffix = "/.default";
         private readonly string _type = typeof(ActiveDirectoryAuthenticationProvider).Name;
         private readonly SqlClientLogger _logger = new SqlClientLogger();
+        private Func<DeviceCodeResult, Task> _deviceCodeFlowCallback;
+
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ctor/*'/>
+        public ActiveDirectoryAuthenticationProvider() => new ActiveDirectoryAuthenticationProvider(DefaultDeviceFlowCallback);
+
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ctor2/*'/>
+        public ActiveDirectoryAuthenticationProvider(Func<DeviceCodeResult, Task> deviceCodeFlowCallbackMethod)
```

Comment:
Opening the up as a public API is very interesting. It would allow us to do some other things, too:

1. For the Interactive flow, we could allow applications to pass in an IWin32Window which we could send to MSAL (.NET Framework only) via .WithParentActivityOrWindow (someone just asked me for this privately) for a more controlled application experience. We could even take in a Microsoft.Identity.Client.Extensibility.ICustomWebUi and pass that via .WithCustomWebUi (netfx/netcore) for even more flexibility.

I'm not sure about adding new public APIs this close to the release without time to discuss. We can always add them later, but it's hard to take them away.

@saurabh500 What do you think about these?

Reply:
I like the idea. I always wondered what would happen for interactive flow different systems. This seems reasonable. 
And +1 to adding a task to backlog and not handling in this release.

Reply:
@cheenamalhotra  do we need to open the whole class up? Wouldn't an API on SqlAuthenticationProvider itself, suffice?

Reply:
`SqlAuthenticationProvider` contains no implementation so there's nothing to override there. Making this class public is the only way to allow users to take advantage of driver's implementation and customize it as needed.

Reply:
This class is meant for inheritance as such. I would recommend only adding APIs which allow us to override the behavior which we want users to override. Adding an API on SqlAuthenticationProvider to set the DeviceAuthFlowCallback should be propagated all the way to `ActiveDirectoryAuthenticationProvider` and that would help achieve what we want. 
In case we open up the whole class then we have to provide a lot of guidance on what it means to override a particular method and this class is based on a lot of internal workings of AAD Auth and whether we are using one kind of Auth over another etc.

Reply:
`SqlAuthenticationProvider` is abstract class, it can't be instantiated by itself.

Users need to be able to instantiate our implementation of `ActiveDirectoryAuthenticationProvider` to override our callback method and for creating instance, it needs to be public.

Reply:
You can always expose a static method in `SqlAuthenticationProvider`. I can imaging different callback per connection, but it doesn't seem like something needed per AuthProvider instance. 
Adding a static to `SqlAuthenticationProvider` is an easier adoption path when it comes to using it.

Reply:
I don't imagine an application instance providing different ways of surfacing the URL for device flow on different connections. This is going to be a console app in most cases. I think a static global callback should be sufficient in this case.

Reply:
@saurabh500 Cheena's point about opening up `ActiveDirectoryAuthenticationProvider` versus `SqlAuthenticationProvider` is that if we don't open up `ActiveDirectoryAuthenticationProvider`, users have to implement all of `ActiveDirectoryAuthenticationProvider` themselves just to customize a few MSAL options. Letting them pass something to handle the device code flow output or interactive window would be much simpler for them.

Reply:
Do you mean we should push this callback to `SqlAuthenticationProviderManager` from static APIs to control custom callback method? But putting callback in `SqlAuthenticationProvider` would mean MSAL.NET using will be part of `SqlAuthenticationProvider` as well as `SqlAuthenticationProviderManager`.

Because this callback is very specific to MSAL.NET as it contains DeviceCodeResult in params, , I don't think MSAL.NET should be used outside the scope of `ActiveDirectoryAuthenticationProvider` that does all the authentication work. So that users are not be forced with using MSAL.NET when implementing `SqlAuthenticationProvider`.

Reply:
I see. I understand better now. In that case it makes sense to open up this type and I guess we will have to add more constructors or properties to accept the customizations. Can we make this class sealed though? I am still of the opinion that this class is not.meant to be inherited

Reply:
Yes sealing it would be ideal 👍